mono/metadata/threads.c

```diff
@@ -4515,61 +4545,93 @@ mono_thread_execute_interruption (void)
 	mono_thread_info_clear_self_interrupt ();
 
 	/* If there's a pending exception and an AbortRequested, the pending exception takes precedence */
-	if (sys_thread->pending_exception) {
-		MonoException *exc;
-
-		exc = sys_thread->pending_exception;
-		sys_thread->pending_exception = NULL;
-
-		UNLOCK_THREAD (thread);
-		return exc;
-	} else if (thread->state & (ThreadState_AbortRequested)) {
-		UNLOCK_THREAD (thread);
-		g_assert (sys_thread->pending_exception == NULL);
-		if (thread->abort_exc == NULL) {
-			/* 
-			 * This might be racy, but it has to be called outside the lock
-			 * since it calls managed code.
-			 */
-			MONO_OBJECT_SETREF (thread, abort_exc, mono_get_exception_thread_abort ());
+	//pending_exception = sys_thread->pending_exception;
+	pending_exception = MONO_HANDLE_NEW_GET (MonoException, sys_thread, pending_exception);
+	if (!MONO_HANDLE_IS_NULL (pending_exception)) {
+		exc = pending_exception;
+		// sys_thread->pending_exception = NULL;
+		MONO_HANDLE_SETRAW (sys_thread, pending_exception, NULL);
+		goto exit;
+	} else if (thread->state & ThreadState_AbortRequested) {
+		g_assert (MONO_HANDLE_IS_NULL (pending_exception));
+
+		// Does the thread already have an abort exception?
+		// If not, create a new one and set it on demand.
+
+		// abort_exc = thread->abort_exc;
+		abort_exc = MONO_HANDLE_NEW_GET (MonoException, thread_handle, abort_exc);
+		if (MONO_HANDLE_IS_NULL (abort_exc)) {
+			// Exception creation has to be outside the lock
+			// since it calls managed code. To avoid race,
+			// drop lock, create local, retry.
+			if (have_new_abort_exc) {
+				// thread->abort_exc = new_abort_exc;
+				MONO_HANDLE_SET (thread_handle, abort_exc, new_abort_exc);
+			} else {
+				UNLOCK_THREAD (thread);
+				unlock = FALSE;
+				new_abort_exc = mono_new_exception_thread_abort ();
+				have_new_abort_exc = TRUE;
+				goto retry;
+			}
 		}
-		return thread->abort_exc;
-	} else if (thread->state & (ThreadState_SuspendRequested)) {
+		// Now the thread has an abort exception to return, because
+		// it already did, or because we just created and set it.
+		// exc = thread->abort_exc; 
+		MONO_HANDLE_GET (exc, thread_handle, abort_exc);
```

Comment:
If you define `MonoExceptionHandle exc = { 0 }`, this line is an error.  (`MONO_HANDLE_GET` does not create a new handle - it assigns to an existing one, which in this case hasn't been allocated yet).
Your options are to allocate an initially-NULL handle when declaring `exc` or to use `MONO_HANDLE_NEW_GET` here. I prefer the former.